Bug 1814993 - Add inactive tabs to the composified `TabsTray`

MozillaNoah opened the pull request:
Changes in this PR:
- Refactored away `TabGrid` +`TabList` in favor of just a `TabList` with a boolean to display the tabs as a grid.
- Added an optional `header` Composable to `TabList` so the `InactiveTabList` Composable can be assimilated into the Lazy grid/list.

The preview for `TabsTray` with the `InactiveTabList` is completely functional! Its basic business logic functionality is plumbed-in using stores + remember variables.



https://user-images.githubusercontent.com/87384386/228334596-98005421-07d7-4a4b-abf7-74f64d059503.mov



### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.








### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1814993

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt:
```diff
@@ -664,6 +680,17 @@ class TabsTrayFragment : AppCompatDialogFragment() {
+            isDisplayedWithBrowserToolbar = true,
+        ).setText(text)
+        snackbar.view.elevation = ELEVATION
+        snackbar.show()
```
I'm going to open a ticket to port this to Compose if this PR is approved

gabrielluong replied:
We used to have a Snackbar implementation that @MatthewTighe did for wallpapers https://github.com/mozilla-mobile/fenix/pull/23196, but I believe it got removed since it wasn't used after further changes to wallpapers as part of MR2. If we were to port this again, I think we should also consider https://bugzilla.mozilla.org/show_bug.cgi?id=1814198. Perhaps utilize the same ticket if it makes sense to, you could reset the assignee since I am not actively working on it, but just assigned myself due to my general interest in it and to help me easily find it. 

My understanding of snackbars is that it is shown at the app level so if the tab trays fragment got dismissed, the snackbar should still be displayed for the set duration. That said I think introducing a composable snackbar does require some more thoughts. It might make sense to keep using `FenixSnackbar` in the meantime until we have a comprehensive understanding of things

MozillaNoah replied:
Appreciate your thoughts here. I'll hold off for now then.

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt:
```diff
@@ -232,6 +235,19 @@ class TabsTrayFragment : AppCompatDialogFragment() {
                         onTabLongClick = tabsTrayInteractor::onTabLongClicked,
+                        onInactiveTabsHeaderClick = tabsTrayInteractor::onInactiveTabsHeaderClicked,
+                        onDeleteAllInactiveTabsClick = tabsTrayInteractor::onDeleteAllInactiveTabsClicked,
+                        onInactiveTabsAutoCloseDialogShown = {
+                            TabsTray.autoCloseSeen.record(NoExtras())
+                        },
```
we may want to refactor this later to the interactor/controller

MatthewTighe replied:
Or into a middleware 😉

MozillaNoah replied:
Opened Bug 1826152

MatthewTighe commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTray.kt:
```diff
@@ -97,42 +128,57 @@ fun TabsTray(
+                        if (inactiveTabs.isNotEmpty()) {
+                            optionalInactiveTabsHeader = {
+                                InactiveTabsList(
+                                    inactiveTabs = inactiveTabs,
+                                    expanded = inactiveTabsExpanded,
+                                    showAutoCloseDialog = showAutoCloseDialog,
+                                    onHeaderClick = onInactiveTabsHeaderClick,
+                                    onDeleteAllButtonClick = onDeleteAllInactiveTabsClick,
+                                    onAutoCloseDismissClick = {
+                                        onInactiveTabAutoCloseDialogCloseButtonClick()
+                                        showAutoCloseDialog = !showAutoCloseDialog
+                                    },
+                                    onEnableAutoCloseClick = {
+                                        onEnableInactiveTabAutoCloseClick()
+                                        showAutoCloseDialog = !showAutoCloseDialog
+                                    },
+                                    onTabClick = onInactiveTabClick,
+                                    onTabCloseClick = onInactiveTabClose,
+                                )
+                            }
+
+                            if (showInactiveTabsDialog) {
+                                onInactiveTabsAutoCloseDialogShown()
+                            }
                         } else {
-                            TabList(
-                                tabs = normalTabs,
-                                onTabClose = onTabClose,
-                                onTabMediaClick = onTabMediaClick,
-                                onTabClick = handleTabClick,
-                                onTabLongClick = onTabLongClick,
-                            )
+                            optionalInactiveTabsHeader = null
```
nit: Switching these cases so that the short null case is first might be it a little easier to read

MatthewTighe commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTray.kt:
```diff
@@ -97,42 +128,57 @@ fun TabsTray(
-                                onTabClick = handleTabClick,
-                                onTabLongClick = onTabLongClick,
-                            )
+                        val showInactiveTabsDialog = showInactiveTabsAutoCloseDialog(inactiveTabs.size)
+                        var showAutoCloseDialog by remember { mutableStateOf(showInactiveTabsDialog) }
```
This function is starting to feel a little crowded, and I'm wondering if it would make sense/be possible to move this dialog logic into `InactiveTabsList`. It might allow us to shrink some of these long parameter lists too, but I don't know if this state needs to memoized at this level for this to work

MozillaNoah replied:
I'll look into reorganizing it, now that I've figured out a good pattern to use stores + Compose.

MozillaNoah replied:
so the main trade off I can foresee is that by ditching the parameters from `InactiveTabList` and moving the logic out of `InactiveTabsViewHolder` and `TabsTray`, is that two stores need to be passed-down. I think this will be inevitable if we ditch interactors, but it also requires the previews to instantiate these large objects.

Not a deal breaker by any means, but just some quick thoughts. I'll have to spin off a different ticket to handle this refactor though.

MatthewTighe replied:
Would it make sense to pass the entire Composable in? something like:

```
fun TabsTray(
    ...
    inactiveTabsContent: @Composable () -> Unit
)
```

There's lots of examples in Compose of this kind of "slot API" like in `Button`, `Scaffold`, etc. I think the question is whether this would be an appropriate use case for that kind of thing

MatthewTighe replied:
We could potentially do this for the banner as well, and help alleviate the concerns you shared above about the parameter list getting even larger.

MozillaNoah replied:
I think it'd be strange to include a Composable as a lambda parameter in this situation, when it'd only be serving the purpose of abstracting out some interactor lambdas, since `TabsTray` won't be used anywhere else in Fenix.

I think it's overall a larger architectural problem to solve of interactors + stores + data flow.

What do we pass for responding to user actions?

Do we only have one store to pass in? Do we sacrifice one degree of testing, if the actions are sent directly from the Composable?

How much data do we pass down to each node of the Composable? Just the raw data it needs? An entire store?

MozillaNoah replied:
Actually, I'd be greatly interested in revisiting this if the [interactor RFC](https://github.com/mozilla-mobile/firefox-android/pull/1466) becomes a thing. It would greatly simplify the lambdas that have to be passed-in, as well as reduce the business logic duplication inside of the previews.

MatthewTighe commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayTabLayouts.kt:
```diff
@@ -34,85 +34,86 @@ import org.mozilla.fenix.theme.FirefoxTheme
-                onLongClick = onTabLongClick,
-            )
-        }
+    if (displayTabsInGrid) {
```
It still kinda feels to me like these separate cases should be separate file-private Composables. If it's a PITA though I think we can just leave as is

MozillaNoah replied:
It's not biggie to separate them back out.

Are you thinking like:

```
 fun TabLayout() {
    if (displayTabsInGrid) {
       TabGrid()
    } else {
        TabList()
    }
}

private fun TabList() 

private fun TabGrid() 

```

Still keeping the root conditional node with private leafs for the actual layouts?

MatthewTighe replied:
Yeah, that's what I'd been thinking!

MatthewTighe reviewed (commented):
Throwing a few comments in here for now. What's a good way to try out the inactive tabs dialog workflows?

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt:
```diff
@@ -232,6 +235,19 @@ class TabsTrayFragment : AppCompatDialogFragment() {
+                        onInactiveTabAutoCloseDialogCloseButtonClick =
+                        tabsTrayInteractor::onAutoCloseDialogCloseButtonClicked,
+                        onEnableInactiveTabAutoCloseClick = {
+                            tabsTrayInteractor::onEnableAutoCloseClicked
```
🤦‍♀️ 🤦‍♀️ 🤦‍♀️  Nope lol. I'll fix this; thanks for catching it.

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTray.kt:
```diff
@@ -38,25 +46,48 @@ import org.mozilla.fenix.theme.FirefoxTheme
 @Suppress("LongMethod", "LongParameterList")
 @Composable
 fun TabsTray(
+    appStore: AppStore,
     tabsTrayStore: TabsTrayStore,
     displayTabsInGrid: Boolean,
+    showInactiveTabsAutoCloseDialog: (Int) -> Boolean,
     onTabClose: (TabSessionState) -> Unit,
     onTabMediaClick: (TabSessionState) -> Unit,
     onTabClick: (TabSessionState) -> Unit,
     onTabMultiSelectClick: (TabSessionState) -> Unit,
     onTabLongClick: (TabSessionState) -> Unit,
+    onInactiveTabsHeaderClick: (Boolean) -> Unit,
+    onDeleteAllInactiveTabsClick: () -> Unit,
+    onInactiveTabsAutoCloseDialogShown: () -> Unit,
+    onInactiveTabAutoCloseDialogCloseButtonClick: () -> Unit,
+    onEnableInactiveTabAutoCloseClick: () -> Unit,
+    onInactiveTabClick: (TabSessionState) -> Unit,
+    onInactiveTabClose: (TabSessionState) -> Unit,
```
Yeah, and we haven't even done the lambdas for the banners yet 😓. Moving away from the lambdas + interactors could certainly help.

MozillaNoah commented:
> What's a good way to try out the inactive tabs dialog workflows?

The inactive tabs dialog shows when you have 20+ inactive tabs, you don't already have auto-close on, and you haven't dismissed the dialog. So force a bunch of tabs inactive (DM if you haven't done this before).

MozillaNoah commented:
Just pushed changes to:

- Split out the grid/list composables into their own functions again.
- Overhauled the preview for `TabsTray`, so it's one parameterized Composable that the actual previews call into.
- Fixed the lambda call for `onEnableAutoCloseClicked`
- Rebase with main

MatthewTighe commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt:
```diff
@@ -219,8 +219,11 @@ class TabsTrayFragment : AppCompatDialogFragment() {
                         tabsTrayStore = tabsTrayStore,
                         displayTabsInGrid = requireContext().settings().gridTabView,
+                        shouldShowInactiveTabsAutoCloseDialog =
+                        requireContext().settings()::shouldShowInactiveTabsAutoCloseDialog,
```
nit: It looks like there's a tab missing here

MozillaNoah replied:
funnily enough, detekt complained about the tab when I originally had it indented here (and with the subsequent comment) 🤷‍♀️

MatthewTighe commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt:
```diff
@@ -232,6 +235,19 @@ class TabsTrayFragment : AppCompatDialogFragment() {
+                            TabsTray.autoCloseSeen.record(NoExtras())
+                        },
+                        onInactiveTabAutoCloseDialogCloseButtonClick =
+                        tabsTrayInteractor::onAutoCloseDialogCloseButtonClicked,
```
nit: tab here too

MatthewTighe commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTray.kt:
```diff
@@ -97,42 +129,56 @@ fun TabsTray(
+                            onTabMediaClick = onTabMediaClick,
+                            onTabClick = handleTabClick,
+                            onTabLongClick = onTabLongClick,
+                            header = optionalInactiveTabsHeader,
```
Ah I hadn't really put it together, but this is `header` parameter is basically what I was suggesting we pass down to the `TabsTray` Composable. `optionalInactiveTabsHeader` would become a parameter instead being put together in this function. Just pointing it out though, I'm still cool with doing it as an optional follow-up later. I suppose it ultimately doesn't matter, but it does feel like that would match up with the pattern of keeping Composables dumb and passing in state/actions.

MozillaNoah replied:
yeah, there's definitely some Goldilocks to be found here, where Composables that are coupled from a feature perspective only get the callbacks/data flow they need and aren't necessarily all coupled to the implementation of a top-level composable.

MatthewTighe reviewed (approved):
couple nits but LGTM otherwise
